arangod/Aql/IResearchViewExecutor.cpp

```diff
@@ -1376,11 +1415,31 @@ size_t IResearchViewMergeExecutor<ordered, materializeType>::skipAll() {
   TRI_ASSERT(this->_filter != nullptr);
 
   size_t skipped = 0;
-
-  while (_heap_it.next()) {
-    skipped++;
+  if (_heap_it.size()) {
+    for (auto& segment : _segments) {
+      TRI_ASSERT(segment.docs);
+      if (filterConditionIsEmpty(&this->infos().filterCondition())) {
+        TRI_ASSERT(segment.segmentIndex < this->_reader->size());
+        auto const live_docs_count =  (*this->_reader)[segment.segmentIndex].live_docs_count();
+        TRI_ASSERT(segment.segmentPos <= live_docs_count);
+        skipped += live_docs_count - segment.segmentPos;
+        segment.segmentPos = live_docs_count;
+      } else {
+       skipped +=  calculateSkipAllCount(this->infos().countApproximate(),
+                                         segment.segmentPos, segment.docs.get());
+      }
+    }
+    // Adjusting by count of docs already consumed by heap but not consumed by executor.
+    // This count is heap size minus 1 already consumed by executor after heap.next.
+    // But we should adjust by the heap size only if the heap was advanced at least once
+    // (heap is actually filled on first next) or we have nothing consumed from doc iterators!
+    if (!_segments.empty() && this->infos().countApproximate() == CountApproximate::Exact &&
+        !filterConditionIsEmpty(&this->infos().filterCondition()) &&
```

Comment:
I suggest we store the result of `filterConditionIsEmpty(&this->infos().filterCondition())` as a member in `infos`